headless-services/spring-boot-language-server/src/main/java/org/springframework/ide/vscode/boot/java/annotations/AnnotationAttributeCompletionProcessor.java

```diff
@@ -114,34 +115,52 @@ else if (node instanceof ArrayInitializer && node.getParent() instanceof Annotat
 	/**
 	 * create the concrete completion proposal
 	 */
-	private void createCompletionProposals(IJavaProject project, TextDocument doc, ASTNode node, String attributeName, Collection<ICompletionProposal> completions, int startOffset, int endOffset,
-			String filterPrefix, Function<String, String> createReplacementText) {
+	private void createCompletionProposals(IJavaProject project, TextDocument doc, ASTNode node, String attributeName,
+			Collection<ICompletionProposal> completions, int startOffset, int endOffset, String filterPrefix,
+			Function<String, String> createReplacementText) {
 
 		Set<String> alreadyMentionedValues = alreadyMentionedValues(node);
 
 		AnnotationAttributeCompletionProvider completionProvider = this.completionProviders.get(attributeName);
 		if (completionProvider != null) {
 			List<String> candidates = completionProvider.getCompletionCandidates(project);
 
-			List<String> filteredCandidates = candidates.stream()
+			if (completionProvider instanceof CronExpressionCompletionProvider) {
```

Comment:
Hmm... this bit does not look pretty... Some Cron completions specific bit in the generic `AnnotationAttributeCompletionProcessor`... Any chance we can push this logic down to the CRON completion provider somehow?

Reply:
I would suggest that we just switch the method `AnnotationAttributeCompletionProvider.getCompletionCandidates` to return labels for the proposals as well and don't add an additional method for that. The `AnnotationAttributeCompletionProcessor` would then not need to distinguish between provider with and without labels and assume instead that the labels are coming from the providers.

Then, we need to adapt all the various provider implementations to return labels as well, but that should't be a big deal.

Reply:
@martinlippert I can do the changes in this PR if it can wait until tomorrow.